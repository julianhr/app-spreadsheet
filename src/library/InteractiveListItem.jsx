/* @jsx jsx */
import React from 'react'
import PropTypes from 'prop-types'
import { jsx, css } from '@emotion/core' // eslint-disable-line
import styled from '@emotion/styled'


const NoOp = () => {}

const Li = styled.li`
  outline: none;
`

function DropdownItem({ item, index, isActive, onMouseEnter }) {
  return (
    <Li
      tabIndex={0}
      onMouseEnter={() => onMouseEnter(index)}
    >
      {isActive && item.active ? item.active : item.base}
    </Li>
  )
}

DropdownItem.propTypes = {
  item: PropTypes.shape({
    base: PropTypes.node.isRequired,
    active: PropTypes.node,
  }),
  index: PropTypes.number.isRequired,
  isActive: PropTypes.bool.isRequired,
  onMouseEnter: PropTypes.func,
}

DropdownItem.defaultProps = {
  onMouseEnter: NoOp,
}

export default DropdownItem
