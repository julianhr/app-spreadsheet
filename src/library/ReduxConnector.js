import appStore from '~/reducers/'


class ReduxConnector {
  constructor(store=appStore) {
    this.store = store
    this.funcs = {}
  }

  get fn() {
    return this.funcs
  }

  getState() {
    return this.store.getState()
  }

  registerFunctions(funcs) {
    for (let [name, fn] of Object.entries(funcs)) {
      this.funcs[name] = async function(...args) {
        await this.store.dispatch( fn(...args) )
      }.bind(this)
    }
  }
}

export default ReduxConnector
