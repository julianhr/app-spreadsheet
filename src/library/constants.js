export const DEFAULT_COL_WIDTH = 140
export const MIN_COL_WIDTH = 12
export const MAX_COL_WIDTH = 1000

