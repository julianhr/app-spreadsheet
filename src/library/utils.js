const BASE_CODE_POINT = 'A'.codePointAt(0)

export function buildUrl(baseUrl, query={}) {
  const urlQuery = Object.entries(query).map(([key, val]) => `${key}=${val}`).join('&')
  return [baseUrl.trim(), urlQuery].join('?')
}

export function clamp(value, min, max) {
  return Math.min( Math.max(value, min), max )
}

export function getRangeArray(min, max, isString=false) {
  const range = []

  for (let num = min; num <= max; num++) {
    range.push( isString ? num.toString() : num )
  }

  return range
}

export const sleep = (ms) => new Promise(resolve => setTimeout(() => resolve(), ms))

export function getColumnNames(cols) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let names = []
  let prevNames = ['']
  let remainCols = cols

  while (remainCols > 0) {
    const newNames = []

    outerLoop:
    for (let colName of prevNames) {  
      for (let letter of alphabet) {
        if (remainCols === 0)  { break outerLoop }
        newNames.push(`${colName}${letter}`)
        remainCols -= 1
      }
    }

    prevNames = newNames
    names = [...names, ...newNames]
  }

  return names
}

export function getCellNames(rows, cols) {
  const names = []
  let colNames = getColumnNames(cols)

  for (let row=1; row <= rows; row++) {
    for (let col of colNames) {
      names.push(`${col}-${row}`)
    }
  }

  return names
}

export function getColumnPosition(colLabel) {
  const letters = colLabel.split('')
  const oneBase = letters.reduce((prev, letter) => (
    prev * 26 + letter.charCodeAt(0) - BASE_CODE_POINT + 1
  ), 0)

  return oneBase - 1
}

export function getColumnLabel(num) {
  const label = []
  const basePoint = BASE_CODE_POINT
  let curr = num + 1

  do {
    curr -= 1
    const letterPosition = curr % 26
    const codePoint = basePoint + letterPosition
    const letter = String.fromCodePoint(codePoint)
    label.push(letter)

    curr = Math.floor(curr / 26)
  } while (curr > 0)

  return label.reverse().join('')
}

export function parseLocation(location) {
  let [col, row] = location.split('-')
  const colIndex = getColumnPosition(col)
  const rowIndex = +row - 1

  return [colIndex, rowIndex]
}

export function isNumber(numStr) {
  const isPeriodsAndDigits = Boolean(numStr.match(/^[.\d]+$/))
  const periodCount = (numStr.match(/\./g) || []).length

  if (!isPeriodsAndDigits) { return false }
  if (periodCount > 1) { return false }
  if (numStr.length === 1 && numStr[0] === '.') { return false }
  return true
}

