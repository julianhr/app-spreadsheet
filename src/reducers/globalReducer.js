import { createReducer } from 'redux-starter-kit'

export const INITIAL_STATE = {
  activeCell: null,
  rows: 14,
  columns: 6,
}

export default createReducer(INITIAL_STATE, {
  'SET_ACTIVE_CELL': (state, { payload }) => {
    state.activeCell = payload
  },
})
