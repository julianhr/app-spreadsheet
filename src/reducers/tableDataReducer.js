import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_DATA': (state, { payload }) => {
    const { location, data } = payload
    state[location] = data
  },

  'DELETE_CELL_DATA': (state, { payload }) => {
    delete state[payload]
  }
})
