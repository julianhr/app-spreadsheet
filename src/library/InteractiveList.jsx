
import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import InteractiveListItem from './InteractiveListItem'


const Ul = styled.ul`
  position: fixed;
  margin: 0;
  padding: 0;
  list-style: none;
`

class InteractiveList extends React.PureComponent {

  static propsTypes = {
    onMouseEnter: PropTypes.func,
    items: PropTypes.array.isRequired,
    top: PropTypes.number.isRequired,
    left: PropTypes.number.isRequired,
    styles: PropTypes.shape({
      ul: PropTypes.object,
      li: PropTypes.object,
    }),
  }

  renderListItems() {
    return this.props.items
      .map((item, index) => (
        <InteractiveListItem
          key={index}
          index={index}
          item={item.item}
          style={{...this.props.styles.li}}
          isActive={item.isActive}
          onMouseEnter={this.props.onMouseEnter}
        />
      ))
  }

  render() {
    const { items, top, left, styles } = this.props

    if (items.length === 0) { return null }

    return (
      <Ul
        tabIndex={-1}
        style={{ top, left, ...styles.ul }}
      >
        {this.renderListItems()}
      </Ul>
    )
  }
}

export default InteractiveList
