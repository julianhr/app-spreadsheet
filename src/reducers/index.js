import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
