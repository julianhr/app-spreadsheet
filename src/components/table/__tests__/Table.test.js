import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent, cleanup } from '@testing-library/react'

import MockApp from '~/__tests__/__mocks__/MockApp'
import ConnectedTable, { Table } from '../Table'
import { appStoreGen } from '~/reducers'
import { sleep } from '~/library/utils'


const requiredProps = {
  rows: 3,
  columns: 2,
}

const testProps = {
  ...requiredProps,
  activeCell: '[data-location="B-2"]',
}

const createApp = (props) => create(<MockApp><Table {...props} /></MockApp>)

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <ConnectedTable {...props} />
    </MockApp>
  )
  return [wrapper, store]
}

describe('Table', () => {
  afterEach(cleanup)

  describe('props', () => {
    it('does not raise warning if all required props are passed', () => {
      expect(() => createApp(testProps)).not.toThrow()
    })

    it('raises warning if required prop is missing', () => {
      Object.keys(requiredProps).forEach(key => {
        const props = {...testProps}
        delete props[key]
        expect(() => createApp(props)).toThrow()
      })
    })
  })

  describe('#handleTableOnClick', () => {
    it('stops propagation of key down events', () => {
      const onClickFn = jest.fn()
      const Listener = ({ children }) => (<div onClick={onClickFn}>{children}</div>)
  
      render(
        <MockApp>
          <Listener>
            <Table {...testProps} />
          </Listener>
        </MockApp>
      )

      const cell = document.querySelector('[data-location="B-2"]')
      fireEvent.click(cell)
      expect(onClickFn).not.toHaveBeenCalled()
    })
  })

  describe('#handleTableOnKeyDown', () => {
    it('stops propagation of key down events', () => {
      const onKeyDownFn = jest.fn()
      const Listener = ({ children }) => (<div onKeyDown={onKeyDownFn}>{children}</div>)
  
      render(
        <MockApp>
          <Listener>
            <Table {...testProps} />
          </Listener>
        </MockApp>
      )

      const cell = document.querySelector('[data-location="B-2"]')
      fireEvent.keyDown(cell, { key: 'a' })
      expect(onKeyDownFn).not.toHaveBeenCalled()
    })

    it('moves focus using keyboard', async () => {
      renderApp(testProps)
      let cell = document.querySelector('[data-location="B-2"]')
      fireEvent.select(cell)
      expect(document.activeElement).toBe(cell)

      fireEvent.keyDown(cell, { key: 'ArrowUp' })
      await sleep(0)
      cell = document.querySelector('[data-location="B-1"]')
      expect(document.activeElement).toBe(cell)
    })
  })

  describe('#moveFocus', () => {
    it('moves focus using keyboard', async (done) => {
      renderApp(testProps)
      const tests = [
        ['ArrowDown', '[data-location="A-2"]'],
        ['ArrowRight', '[data-location="B-2"]'],
        ['Enter', '[data-location="B-3"]'],
        ['ArrowLeft', '[data-location="A-3"]'],
      ]
      let cell = document.querySelector('[data-location="A-1"]')
      fireEvent.select(cell)
      expect(document.activeElement).toBe(cell)

      for (let [key, location] of tests) {
        fireEvent.keyDown(cell, { key })
        await sleep(0)
        cell = document.querySelector(location)
        expect(document.activeElement).toBe(cell)
      }

      done()
    })
  })

  describe('functional tests', () => {
    it('matches snapshot', () => {
      const store = appStoreGen()
      const { global } = store.getState()
      global.rows = 2
      global.columns = 3
      renderApp(testProps, store)
      const el = document.querySelector('[data-table="app"]')
      expect(el).toMatchSnapshot()
    })
  })
})
