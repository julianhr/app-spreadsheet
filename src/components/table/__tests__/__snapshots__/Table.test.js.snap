// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table functional tests matches snapshot 1`] = `
<div
  class="css-1pfmi74-Grid e1tmjcr20"
  data-table="app"
>
  <div
    class="css-g0yt5d-Row e40r6tf0"
    data-row="header"
  >
    <div
      class="row-label-width col-label-height css-ql6bmv-InnerBorder e40r6tf1"
    />
    <div
      class="col-label-height css-va29v5-InnerBorder-ColHeader e11zdtao0"
      data-col="A"
    >
      A
      <button
        class="css-1oguur2-BtnWidth e11zdtao1"
        tabindex="-1"
      />
    </div>
    <div
      class="col-label-height css-va29v5-InnerBorder-ColHeader e11zdtao0"
      data-col="B"
    >
      B
      <button
        class="css-1oguur2-BtnWidth e11zdtao1"
        tabindex="-1"
      />
    </div>
    <div
      class="col-label-height css-va29v5-InnerBorder-ColHeader e11zdtao0"
      data-col="C"
    >
      C
      <button
        class="css-1oguur2-BtnWidth e11zdtao1"
        tabindex="-1"
      />
    </div>
  </div>
  <div
    class="css-g0yt5d-Row exk1gay0"
    data-row="1"
  >
    <div
      class="row-label-width row-label-height css-1cj8orw-Wrapper e1d5mcq00"
      data-row="1"
    >
      1
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="A-1"
        tabindex="0"
      />
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="B-1"
        tabindex="0"
      />
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="C-1"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="css-g0yt5d-Row exk1gay0"
    data-row="2"
  >
    <div
      class="row-label-width row-label-height css-1cj8orw-Wrapper e1d5mcq00"
      data-row="2"
    >
      2
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="A-2"
        tabindex="0"
      />
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="B-2"
        tabindex="0"
      />
    </div>
    <div
      class="row-label-height css-1p2iaz6-Wrapper-ResultCell egobumn0"
    >
      <div
        class="css-17w7ktc-Cell egobumn1"
        data-cell="result"
        data-location="C-2"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
